File: clients/flutter/chex_parity.dart
// Checks both Flutter backends against the real `chex` binary (the oracle).
//
// chex_ffi.dart calls the core, so what is being checked there is the binding —
// request marshalling, allocation, error decoding. chex_web.dart is a genuine
// reimplementation, so for it this is the lockstep guarantee it depends on.
//
//   CHEX_LIBRARY=target/release/libchex.dylib \
//     dart run clients/flutter/chex_parity.dart ./target/release/chex
//
// Exits 0 when every case matches, 1 otherwise. Uses dart:io only in this
// harness; the backends under test do not.

import 'dart:convert';
import 'dart:io';

import 'chex_error.dart';
import 'chex_ffi.dart' as ffi;
import 'chex_web.dart' as web;

class Case {
  final String name;
  final Map<String, dynamic> schema;
  final Map<String, dynamic> data;
  final bool valid;
  const Case(this.name, this.schema, this.data, this.valid);
}

const cases = <Case>[
  Case('primitive pass', {'age': r'^[0-9]+$'}, {'age': 30}, true),
  Case('primitive fail', {'age': r'^[0-9]+$'}, {'age': 'x'}, false),
  Case('boolean coercion', {'active': r'^(true|false)$'}, {'active': true}, true),
  Case('nullable absent', {'nickname?': r'^[a-z]+$'}, {}, true),
  Case('nullable present ok', {'nickname?': r'^[a-z]+$'}, {'nickname': 'ada'}, true),
  Case('nullable present bad', {'nickname?': r'^[a-z]+$'}, {'nickname': 'A1'}, false),
  Case('missing required', {'age': r'^[0-9]+$'}, {}, false),
  Case('unknown property', {'age': r'^[0-9]+$'}, {'age': 1, 'extra': 'x'}, false),
  Case('nested object ok', {'addr': {'city': r'^[A-Za-z]+$'}}, {'addr': {'city': 'Lagos'}}, true),
  Case('nested object bad', {'addr': {'city': r'^[A-Za-z]+$'}}, {'addr': {'city': 'L4'}}, false),
  Case('object type mismatch', {'addr': {'city': r'^[A-Za-z]+$'}}, {'addr': 'x'}, false),
  Case('scalar array ok', {'tags': [r'^[a-z]+$']}, {'tags': ['bun', 'web']}, true),
  Case('scalar array bad', {'tags': [r'^[a-z]+$']}, {'tags': ['bun', 'W1']}, false),
  Case('array type mismatch', {'tags': [r'^[a-z]+$']}, {'tags': 'nope'}, false),
  Case('array of objects ok', {'items': [{'sku': r'^[A-Z0-9-]+$', 'gift?': r'^(true|false)$'}]},
      {'items': [{'sku': 'AB-1'}, {'sku': 'CD-2', 'gift': true}]}, true),
  Case('array of objects bad', {'items': [{'sku': r'^[A-Z0-9-]+$'}]}, {'items': [{'sku': 'ab-1'}]}, false),
  Case('record ok', {'meta': {r'^[a-z_]+$': r'^.+$'}}, {'meta': {'a_b': 'x'}}, true),
  Case('record bad key', {'meta': {r'^[a-z_]+$': r'^.+$'}}, {'meta': {'A': 'x'}}, false),
  Case('record bad value', {'meta': {r'^[a-z]+$': r'^[0-9]+$'}}, {'meta': {'a': 'x'}}, false),
];

typedef Validator = Map<String, dynamic> Function(
    Map<String, dynamic> schema, Map<String, dynamic> data,
    {String label});

bool accepts(Validator validator, Map<String, dynamic> schema, Map<String, dynamic> data) {
  try {
    validator(schema, data);
    return true;
  } on CHEXError {
    return false;
  }
}

int _counter = 0;
bool binaryAccepts(String bin, Directory dir, Map<String, dynamic> schema, Map<String, dynamic> data) {
  final file = File('${dir.path}/s${_counter++}.schema.json')..writeAsStringSync(jsonEncode(schema));
  final result = Process.runSync(bin, ['validate', file.path, jsonEncode(data)]);
  return result.exitCode == 0;
}

void main(List<String> args) {
  final bin = args.isNotEmpty ? args[0] : 'chex';
  final dir = Directory.systemTemp.createTempSync('chex-flutter-');
  var failures = 0;
  try {
    for (final c in cases) {
      final oracle = binaryAccepts(bin, dir, c.schema, c.data);
      final viaFfi = accepts(ffi.validate, c.schema, c.data);
      final viaWeb = accepts(web.validate, c.schema, c.data);
      if (oracle != c.valid || viaFfi != oracle || viaWeb != oracle) {
        failures++;
        stderr.writeln(
            "MISMATCH ${c.name}: expected=${c.valid} binary=$oracle ffi=$viaFfi web=$viaWeb");
      }
    }
  } finally {
    dir.deleteSync(recursive: true);
  }
  if (failures == 0) {
    stdout.writeln('parity OK: ${cases.length} cases agree with the chex binary');
  } else {
    stderr.writeln('$failures/${cases.length} cases FAILED');
    exit(1);
  }
}

File: clients/ios/ChexParity.swift
// Smoke test for the C ABI binding in Chex.swift.
//
// Chex.swift calls the same engine the binary does, so the validation rules
// can't disagree — what this checks is the binding: request marshalling, the
// result buffer, and error decoding. It runs a battery of cases through both and
// asserts they agree on accept/reject.
//
//   swiftc -import-objc-header include/chex.h \
//     clients/ios/Chex.swift clients/ios/ChexParity.swift \
//     -L target/release -lchex -o /tmp/chex-ios
//   /tmp/chex-ios ./target/release/chex
//
// Exits 0 when every case matches, 1 otherwise. Foundation only.

import Foundation

struct ParityCase {
    let name: String
    let schema: String
    let data: String
    let valid: Bool
}

let cases: [ParityCase] = [
    ParityCase(name: "primitive pass", schema: #"{"age":"^[0-9]+$"}"#, data: #"{"age":30}"#, valid: true),
    ParityCase(name: "primitive fail", schema: #"{"age":"^[0-9]+$"}"#, data: #"{"age":"x"}"#, valid: false),
    ParityCase(name: "boolean coercion", schema: #"{"active":"^(true|false)$"}"#, data: #"{"active":true}"#, valid: true),
    ParityCase(name: "nullable absent", schema: #"{"nickname?":"^[a-z]+$"}"#, data: #"{}"#, valid: true),
    ParityCase(name: "nullable present ok", schema: #"{"nickname?":"^[a-z]+$"}"#, data: #"{"nickname":"ada"}"#, valid: true),
    ParityCase(name: "nullable present bad", schema: #"{"nickname?":"^[a-z]+$"}"#, data: #"{"nickname":"A1"}"#, valid: false),
    ParityCase(name: "missing required", schema: #"{"age":"^[0-9]+$"}"#, data: #"{}"#, valid: false),
    ParityCase(name: "unknown property", schema: #"{"age":"^[0-9]+$"}"#, data: #"{"age":1,"extra":"x"}"#, valid: false),
    ParityCase(name: "nested object ok", schema: #"{"addr":{"city":"^[A-Za-z]+$"}}"#, data: #"{"addr":{"city":"Lagos"}}"#, valid: true),
    ParityCase(name: "nested object bad", schema: #"{"addr":{"city":"^[A-Za-z]+$"}}"#, data: #"{"addr":{"city":"L4"}}"#, valid: false),
    ParityCase(name: "object type mismatch", schema: #"{"addr":{"city":"^[A-Za-z]+$"}}"#, data: #"{"addr":"x"}"#, valid: false),
    ParityCase(name: "scalar array ok", schema: #"{"tags":["^[a-z]+$"]}"#, data: #"{"tags":["bun","web"]}"#, valid: true),
    ParityCase(name: "scalar array bad", schema: #"{"tags":["^[a-z]+$"]}"#, data: #"{"tags":["bun","W1"]}"#, valid: false),
    ParityCase(name: "array type mismatch", schema: #"{"tags":["^[a-z]+$"]}"#, data: #"{"tags":"nope"}"#, valid: false),
    ParityCase(name: "array of objects ok", schema: #"{"items":[{"sku":"^[A-Z0-9-]+$","gift?":"^(true|false)$"}]}"#, data: #"{"items":[{"sku":"AB-1"},{"sku":"CD-2","gift":true}]}"#, valid: true),
    ParityCase(name: "array of objects bad", schema: #"{"items":[{"sku":"^[A-Z0-9-]+$"}]}"#, data: #"{"items":[{"sku":"ab-1"}]}"#, valid: false),
    ParityCase(name: "record ok", schema: #"{"meta":{"^[a-z_]+$":"^.+$"}}"#, data: #"{"meta":{"a_b":"x"}}"#, valid: true),
    ParityCase(name: "record bad key", schema: #"{"meta":{"^[a-z_]+$":"^.+$"}}"#, data: #"{"meta":{"A":"x"}}"#, valid: false),
    ParityCase(name: "record bad value", schema: #"{"meta":{"^[a-z]+$":"^[0-9]+$"}}"#, data: #"{"meta":{"a":"x"}}"#, valid: false),
]

func inProcessAccepts(_ schemaJson: String, _ dataJson: String) -> Bool {
    guard let schema = (try? JSONSerialization.jsonObject(with: Data(schemaJson.utf8))) as? [String: Any],
          let data = (try? JSONSerialization.jsonObject(with: Data(dataJson.utf8))) as? [String: Any] else {
        return false
    }
    do { _ = try CHEXValidator.validate(schema, data); return true } catch { return false }
}

var counter = 0
func binaryAccepts(_ bin: String, _ dir: URL, _ schemaJson: String, _ dataJson: String) -> Bool {
    let file = dir.appendingPathComponent("s\(counter).schema.json")
    counter += 1
    try? schemaJson.write(to: file, atomically: true, encoding: .utf8)
    let proc = Process()
    proc.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    proc.arguments = [bin, "validate", file.path, dataJson]
    proc.standardOutput = Pipe()
    proc.standardError = Pipe()
    do { try proc.run() } catch { return false }
    proc.waitUntilExit()
    return proc.terminationStatus == 0
}

@main
struct ChexParityMain {
    static func main() {
        let bin = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "chex"
        let dir = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("chex-ios-\(UUID().uuidString)")
        try! FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        var failures = 0
        for testCase in cases {
            let oracle = binaryAccepts(bin, dir, testCase.schema, testCase.data)
            let inProc = inProcessAccepts(testCase.schema, testCase.data)
            if !(oracle == testCase.valid && inProc == oracle) {
                failures += 1
                FileHandle.standardError.write("MISMATCH \(testCase.name): expected=\(testCase.valid) binary=\(oracle) inProcess=\(inProc)\n".data(using: .utf8)!)
            }
        }
        try? FileManager.default.removeItem(at: dir)
        if failures == 0 {
            print("parity OK: \(cases.count) cases agree with the chex binary")
        } else {
            FileHandle.standardError.write("\(failures)/\(cases.count) cases FAILED\n".data(using: .utf8)!)
            exit(1)
        }
    }
}
